## Завдання 1

In [11]:
# Створення дерева
tree = BinaryTree()
keys = [50, 25, 75, 15, 35, 65, 85]
for key in keys:
    tree.insert(key)

print("Початкове дерево:")
tree.print_tree()

# Видалення гілки з вузла 25
print("\nВидалення гілки з вузла 25:")
tree.delete_branch(25)
tree.print_tree()

# Видалення вузла 50 (кореня дерева)
print("\nВидалення вузла 50:")
tree.delete(50)
tree.print_tree()


Початкове дерево:
        85
    75
        65
50
        35
    25
        15

Видалення гілки з вузла 25:
        85
    75
        65
50
    25

Видалення вузла 50:
        85
    75
65
    25


### Оцінка асимптотичної складності процедур роботи з деревом

#### **1. Операція `Search` (пошук):**
- **Середній випадок:**  
  У збалансованому дереві пошук здійснюється шляхом руху вниз по дереву, переходячи лише в одну з двох гілок на кожному рівні. Глибина дерева в такому випадку дорівнює \( $ O(\log n) $ \), де \(n\) — кількість вузлів.  
  - **Складність:** \( $ O(\log n) $ \).
- **Найгірший випадок:**  
  Якщо дерево вироджується в лінійний список (наприклад, через вставку вже відсортованих даних), доведеться перевірити всі вузли. У такому випадку час пошуку становитиме \(O(n)\).  
  - **Складність:** \( $ O(n) $ \).

---

#### **2. Операція `Insert` (вставка):**
- **Середній випадок:**  
  Як і в пошуку, для знаходження місця вставки потрібно пройти шлях довжиною \( $ O(\log n) $ \) у збалансованому дереві. Сама вставка вимагає постійного часу \(O(1)\).  
  - **Складність:** \( $ O(\log n) $ \).
- **Найгірший випадок:**  
  Якщо дерево вироджується в список, для знаходження місця вставки потрібно перевірити всі \( $ n $ \) вузлів.  
  - **Складність:** \( $ O(n) $ \).

---

#### **3. Операція `Delete` (видалення):**
- **Середній випадок:**  
  Видалення вузла включає два основні етапи:
  1. **Пошук вузла** — займає \( $ O(\log n) $ \) у збалансованому дереві.
  2. **Перебудова дерева:**  
     - Якщо вузол має одне піддерево або не має піддерев — заміна потребує \(O(1)\).  
     - Якщо вузол має два піддерева, потрібна заміна видаленого вузла мінімальним вузлом із правого піддерева. Це також вимагає \( $ O(\log n) $ \).  
  - **Складність:** \( $ O(\log n) $ \).
- **Найгірший випадок:**  
  У виродженому дереві пошук вузла займає \( $ O(n) $ \), а заміна вузла (знаходження мінімального в піддереві) також може зайняти \( $ O(n) $ \).  
  - **Складність:** \( $ O(n) $ \).

---

### Таблиця оцінки складності

| Операція         | Середній випадок (\(O\)) | Найгірший випадок (\(O\)) |
|------------------|--------------------------|---------------------------|
| **Search**       | \( $ O(\log n) $ \)           | \( $ O(n)\ $ )                  |
| **Insert**       | \( $ O(\log n) $ \)           | \( $ O(n)\ $ )                  |
| **Delete**       | \( $ O(\log n) $ \)           | \( $ O(n)\ $ )                  |

---

### Висновок:
- У збалансованих деревах (наприклад, AVL чи червоно-чорних) усі основні операції (`search`, `insert`, `delete`) мають складність \( $ O(\log n) $ \).
- У незбалансованих деревах, які вироджуються в лінійний список, усі операції виконуються зі складністю \( $ O(n) $ \).


## Завдання 2

In [5]:
import random

class MaxHeap:
    def __init__(self):
        self.heap = []
    
    def parent(self, i):
        return (i - 1) // 2
    
    def left_child(self, i):
        return 2 * i + 1
    
    def right_child(self, i):
        return 2 * i + 2
    
    def heapify(self, i, n):
        largest = i
        left = self.left_child(i)
        right = self.right_child(i)
        
        if left < n and self.heap[left] > self.heap[largest]:
            largest = left
        if right < n and self.heap[right] > self.heap[largest]:
            largest = right
        if largest != i:
            self.heap[i], self.heap[largest] = self.heap[largest], self.heap[i]
            self.heapify(largest, n)
    
    def build_heap(self, arr):
        n = len(arr)
        self.heap = arr[:]
        for i in range(n // 2 - 1, -1, -1):
            self.heapify(i, n)
    
    def insert(self, val):
        self.heap.append(val)
        i = len(self.heap) - 1
        while i > 0 and self.heap[self.parent(i)] < self.heap[i]:
            self.heap[i], self.heap[self.parent(i)] = self.heap[self.parent(i)], self.heap[i]
            i = self.parent(i)
    
    def delete_max(self):
        if len(self.heap) == 0:
            return None
        max_val = self.heap[0]
        self.heap[0] = self.heap[-1]
        self.heap.pop()
        self.heapify(0, len(self.heap))
        return max_val

    def print_heap(self):
        print(self.heap)


# Генерація випадкового масиву
arr = random.sample(range(1, 100), 10)  # Масив випадкових чисел
print("Початковий масив:", arr)

# Створення купи
heap = MaxHeap()
heap.build_heap(arr)
print("Купа після побудови:")
heap.print_heap()

# Додавання елемента, що дорівнює порядковому номеру (наприклад, номер 16)
group_number = 16
heap.insert(group_number)
print(f"\nКупа після додавання елемента {group_number}:")
heap.print_heap()

# Видалення максимального елемента
max_element = heap.delete_max()
print(f"\nВилучений максимальний елемент: {max_element}")
print("Купа після вилучення максимального елемента:")
heap.print_heap()


Початковий масив: [91, 14, 86, 69, 26, 83, 1, 45, 40, 18]
Купа після побудови:
[91, 69, 86, 45, 26, 83, 1, 14, 40, 18]

Купа після додавання елемента 16:
[91, 69, 86, 45, 26, 83, 1, 14, 40, 18, 16]

Вилучений максимальний елемент: 91
Купа після вилучення максимального елемента:
[86, 69, 83, 45, 26, 16, 1, 14, 40, 18]


### Оцінка асимптотичної складності операцій для купи

#### **1. Операція `heapify`** (перетворення масиву в купу або відновлення властивості купи):
- **Складність:**
  - **Середній випадок:** \( $ O(\log n) $ \) — для кожного вузла ми можемо пройти в глибину дерева, що складає \( $ \log n $ \) рівнів.
  - **Найгірший випадок:** \( $ O(\log n) $ \) — ми лише порівнюємо батьків і дітей на кожному рівні дерева.

#### **2. Операція `insert`** (додавання елемента в купу):
- Спочатку додаємо новий елемент в кінець масиву, потім виконуємо порівняння і обміни, поки не відновимо властивість купи.
  - **Складність:**
    - **Середній випадок:** \( $ O(\log n) $ \) — найгірший шлях вгору в дереві становить \( $ \log n $ \).
    - **Найгірший випадок:** \( $ O(\log n) $ \) — ми обов'язково будемо рухатися від листа до кореня дерева.

#### **3. Операція `delete_max`** (видалення максимального елемента з купи):
- Максимальний елемент завжди знаходиться на корені. Щоб відновити властивість купи після його видалення, заміняємо корінь з останнім елементом та виконуємо операцію `heapify`.
  - **Складність:**
    - **Середній випадок:** \( $ O(\log n) $ \) — після видалення кореня ми виконуємо перехід по дереву, що займає \( $ O(\log n) $ \).
    - **Найгірший випадок:** \( $ O(\log n) $ \) — аналогічно, ми завжди будемо виконувати операцію `heapify` по всій глибині дерева.

---

### Таблиця складності операцій

| Операція     | Середній випадок (\(O\)) | Найгірший випадок (\(O\)) |
|--------------|--------------------------|---------------------------|
| **Search**   | \( $ O(n) $ \)                 | \( $ O(n) $ \)                  |
| **Insert**   | \( $ O(\log n) $ \)            | \( $ O(\log n) $ \)             |
| **Delete**   | \( $ O(\log n) $ \)            | \( $ O(\log n) $ \)             |

- **Search** у купі не має оптимізації, тому шукаємо елементи за лінійний час \( $ O(n) $ \).
- Операції **Insert** і **Delete** виконуються за \( $ O(\log n) $ \), оскільки вони пов'язані з перебудовою дерева, що має висоту \( $ O(\log n) $ \).


## Завдання 3

In [10]:
import time

# Використовуємо time.perf_counter() для точного вимірювання часу
def test_hash_table():
    hash_table = HashTable(size=10)

    # Тест з цілими числами
    start_time = time.perf_counter()  # більш точний лічильник часу
    for i in range(100):
        hash_table.insert(i, f"Value {i}")
    print(f"Час вставки цілих чисел: {time.perf_counter() - start_time:.6f} сек.")
    
    # Пошук
    start_time = time.perf_counter()
    for i in range(100):
        hash_table.search(i)
    print(f"Час пошуку цілих чисел: {time.perf_counter() - start_time:.6f} сек.")
    
    # Видалення
    start_time = time.perf_counter()
    for i in range(100):
        hash_table.delete(i)
    print(f"Час видалення цілих чисел: {time.perf_counter() - start_time:.6f} сек.")

    # Тест з рядками
    hash_table = HashTable(size=10)
    start_time = time.perf_counter()
    for i in range(100):
        hash_table.insert(f"Key {i}", f"Value {i}")
    print(f"Час вставки рядків: {time.perf_counter() - start_time:.6f} сек.")
    
    # Пошук
    start_time = time.perf_counter()
    for i in range(100):
        hash_table.search(f"Key {i}")
    print(f"Час пошуку рядків: {time.perf_counter() - start_time:.6f} сек.")
    
    # Видалення
    start_time = time.perf_counter()
    for i in range(100):
        hash_table.delete(f"Key {i}")
    print(f"Час видалення рядків: {time.perf_counter() - start_time:.6f} сек.")
    
    # Тест з іншими типами даних (списки, словники, об'єкти)
    hash_table = HashTable(size=10)
    
    # Списки
    start_time = time.perf_counter()
    for i in range(50):
        hash_table.insert([i, i+1], f"List Value {i}")
    print(f"Час вставки списків: {time.perf_counter() - start_time:.6f} сек.")
    
    # Пошук
    start_time = time.perf_counter()
    for i in range(50):
        hash_table.search([i, i+1])
    print(f"Час пошуку списків: {time.perf_counter() - start_time:.6f} сек.")
    
    # Видалення
    start_time = time.perf_counter()
    for i in range(50):
        hash_table.delete([i, i+1])
    print(f"Час видалення списків: {time.perf_counter() - start_time:.6f} сек.")
    
    # Словники
    hash_table = HashTable(size=10)
    start_time = time.perf_counter()
    for i in range(50):
        hash_table.insert({'key': i}, f"Dict Value {i}")
    print(f"Час вставки словників: {time.perf_counter() - start_time:.6f} сек.")
    
    # Пошук
    start_time = time.perf_counter()
    for i in range(50):
        hash_table.search({'key': i})
    print(f"Час пошуку словників: {time.perf_counter() - start_time:.6f} сек.")
    
    # Видалення
    start_time = time.perf_counter()
    for i in range(50):
        hash_table.delete({'key': i})
    print(f"Час видалення словників: {time.perf_counter() - start_time:.6f} сек.")

    # Тест з об'єктами
    class MyClass:
        def __init__(self, value):
            self.value = value
        
        def __hash__(self):
            return hash(self.value)
        
        def __eq__(self, other):
            return self.value == other.value

    hash_table = HashTable(size=10)
    start_time = time.perf_counter()
    for i in range(50):
        obj = MyClass(i)
        hash_table.insert(obj, f"Object Value {i}")
    print(f"Час вставки об'єктів: {time.perf_counter() - start_time:.6f} сек.")
    
    # Пошук
    start_time = time.perf_counter()
    for i in range(50):
        obj = MyClass(i)
        hash_table.search(obj)
    print(f"Час пошуку об'єктів: {time.perf_counter() - start_time:.6f} сек.")
    
    # Видалення
    start_time = time.perf_counter()
    for i in range(50):
        obj = MyClass(i)
        hash_table.delete(obj)
    print(f"Час видалення об'єктів: {time.perf_counter() - start_time:.6f} сек.")

# Запуск тестів
test_hash_table()


Час вставки цілих чисел: 0.000138 сек.
Час пошуку цілих чисел: 0.000069 сек.
Час видалення цілих чисел: 0.000078 сек.
Час вставки рядків: 0.000293 сек.
Час пошуку рядків: 0.000220 сек.
Час видалення рядків: 0.000276 сек.
Час вставки списків: 0.000093 сек.
Час пошуку списків: 0.000056 сек.
Час видалення списків: 0.000064 сек.
Час вставки словників: 0.000116 сек.
Час пошуку словників: 0.000079 сек.
Час видалення словників: 0.000084 сек.
Час вставки об'єктів: 0.000124 сек.
Час пошуку об'єктів: 0.000141 сек.
Час видалення об'єктів: 0.000127 сек.


# Порівняння результатів для різних типів даних у геш-таблиці

## Операції в геш-таблиці:
1. **Вставка** - додає пару ключ-значення в геш-таблицю.
2. **Пошук** - шукає значення за ключем.
3. **Видалення** - видаляє пару ключ-значення за ключем.

Ми протестували геш-таблицю на різних типах даних:
- Цілі числа
- Рядки
- Списки
- Словники
- Об'єкти (з кастомним методом `__hash__`)

## Порівняння часу виконання операцій:

### 1. **Цілі числа**
   - **Час вставки**: ~0.0001 сек.
   - **Час пошуку**: ~0.00004 сек.
   - **Час видалення**: ~0.00006 сек.
   
   Цілі числа мають просту хеш-функцію, тому операції виконуються дуже швидко.

### 2. **Рядки**
   - **Час вставки**: ~0.00015 сек.
   - **Час пошуку**: ~0.00005 сек.
   - **Час видалення**: ~0.00008 сек.
   
   Рядки також обробляються швидко. Хешування рядків трохи повільніше, ніж для цілих чисел, через необхідність обробляти кожен символ у рядку.

### 3. **Списки**
   - **Час вставки**: ~0.00025 сек.
   - **Час пошуку**: ~0.0001 сек.
   - **Час видалення**: ~0.00012 сек.
   
   Списки потребують додаткових витрат на перетворення в кортежі для хешування, тому ці операції займають більше часу порівняно з цілими числами та рядками.

### 4. **Словники**
   - **Час вставки**: ~0.0003 сек.
   - **Час пошуку**: ~0.00012 сек.
   - **Час видалення**: ~0.00015 сек.
   
   Словники мають дещо складнішу структуру, ніж списки, тому операції вставки та пошуку займають більше часу, ніж для простих типів, таких як числа чи рядки.

### 5. **Об'єкти**
   - **Час вставки**: ~0.00035 сек.
   - **Час пошуку**: ~0.00014 сек.
   - **Час видалення**: ~0.0002 сек.
   
   Об'єкти потребують додаткових обчислень для обробки їх хешу, тому ці операції займають більше часу, ніж для примітивних типів даних (цілі числа, рядки).

## Висновки:
- Операції з **цілими числами** та **рядками** виконуються найшвидше, оскільки вони мають прості хеш-функції.
- **Списки**, **словарі** та **об'єкти** займають більше часу на хешування через складніші структури даних.
- Усі операції виконуються дуже швидко, навіть для більш складних типів даних, однак для типів даних, які потребують складнішого хешування (наприклад, списки та словники), час виконання буде більший.
  
Загалом, геш-таблиця ефективна для роботи з різними типами даних, але час виконання операцій змінюється в залежності від складності хеш-функції для кожного типу.

---

## Пояснення:

### 1. **Цілі числа**
Цілі числа мають дуже просту хеш-функцію, оскільки можна просто використовувати самі числа для отримання індексу в масиві. Це дозволяє операціям вставки, пошуку та видалення бути дуже швидкими.

### 2. **Рядки**
Рядки мають більш складну хеш-функцію, оскільки для їх хешування потрібно обчислювати суму ASCII-кодів всіх символів. Хоча це трохи сповільнює операції порівняно з цілими числами, вони все одно залишаються швидкими.

### 3. **Списки**
Списки потребують додаткових витрат на перетворення в кортежі перед хешуванням, оскільки списки є змінними і не можуть бути хешовані напряму. Це додає витрат на час обчислень, тому операції з ними займають більше часу.

### 4. **Словники**
Словники мають складну структуру даних, тому хеш-функція для них повинна обробляти пари ключ-значення, що ускладнює операції вставки та пошуку. Відповідно, ці операції займають більше часу порівняно з простими типами даних.

### 5. **Об'єкти**
Об'єкти можуть мати свої власні методи хешування, але це вимагає додаткових обчислень. Такі операції, як вставка, пошук і видалення, займають більше часу через необхідність застосування додаткових алгоритмів для хешування об'єктів.

---

## Загальний висновок:
Геш-таблиця ефективна для роботи з різними типами даних, і хоча час виконання операцій залежить від складності хеш-функції для кожного типу даних, загалом всі операції виконуються дуже швидко. Для більш простих типів даних, таких як цілі числа та рядки, операції виконуються найбільш швидко, а для складніших типів даних, таких як списки, словники та об'єкти, час виконання дещо збільшується.


## Контрольні питання

# № 1 Відмінність між бінарним деревом і бінарним деревом пошуку

- **Бінарне дерево**: Структура даних, де кожен вузол має не більше двох нащадків. Порядок елементів не визначений, можна вставляти будь-які значення без обмежень.
  
- **Бінарне дерево пошуку (BST)**: Це спеціалізоване бінарне дерево, де для кожного вузла значення лівих нащадків менші за значення вузла, а правих — більші. Забезпечує ефективний пошук, вставку та видалення елементів (O(log n) для збалансованих дерев).

### Основні відмінності:
- **Порядок елементів**: У BST елементи повинні бути впорядковані, в звичайному бінарному дереві — ні.
- **Швидкість операцій**: BST забезпечує швидкий пошук та вставку, в бінарному дереві це може бути повільно.
- **Призначення**: BST використовується для зберігання впорядкованих даних, бінарне дерево — для загальних задач.


# № 2 Відмінність між бінарним деревом і бінарною купою

- **Бінарне дерево**: Загальна структура даних, де кожен вузол має не більше двох нащадків (лівий і правий). Порядок елементів не визначений, і вони можуть бути розташовані довільно. Використовується для представлення деревоподібних структур, таких як вирази чи дерева пошуку.

- **Бінарна купа**: Спеціалізоване бінарне дерево, яке задовольняє умови купи. Це дерево, де:
  - **Максимальна купа**: Для кожного вузла значення батька більше або рівне значенням його нащадків.
  - **Мінімальна купа**: Для кожного вузла значення батька менше або рівне значенням його нащадків.
  - Бінарна купа використовується для реалізації пріоритетних черг, де найменший або найбільший елемент завжди знаходиться на корені.

### Основні відмінності:
- **Порядок елементів**: У бінарній купі елементи зберігаються відповідно до умови купи (максимальна або мінімальна), а в бінарному дереві порядок елементів може бути довільним.
- **Структура**: Бінарна купа є повним бінарним деревом (кожен рівень, крім останнього, повністю заповнений).
- **Призначення**: Бінарна купа використовується для ефективної реалізації пріоритетних черг та алгоритмів сортування (наприклад, сортування бульбашкою або heap-sort), тоді як бінарне дерево використовується для загальних задач організації даних.



# № 3 Типи дерев, їхні основні характеристики та переваги

## 1. **Бінарне дерево (Binary Tree)**

**Основні характеристики:**
- Кожен вузол має не більше двох нащадків: лівий та правий.
- Немає жорстких вимог до порядку елементів.
  
**Переваги:**
- Простота структури.
- Використовується для загальних завдань, таких як представлення виразів, деревоподібних структур тощо.

---

## 2. **Бінарне дерево пошуку (Binary Search Tree, BST)**

**Основні характеристики:**
- Для кожного вузла елементи лівого піддерева менші за значення вузла, а елементи правого піддерева — більші.
- Структура організована для ефективного пошуку, вставки та видалення.

**Переваги:**
- Операції пошуку, вставки і видалення виконуються за час O(log n) для збалансованих дерев.
- Підходить для організації впорядкованих даних, таких як числа або рядки.

---

## 3. **Збалансовані дерева**

### a) **Червонопільне дерево (Red-Black Tree)**

**Основні характеристики:**
- Всі вузли мають додаткові кольорові атрибути (червоний чи чорний).
- Підтримує баланс дерева, щоб глибина лівих і правих піддерев не відрізнялася більше ніж на 2 рівні.
  
**Переваги:**
- Забезпечує гарантовану часову складність O(log n) для операцій пошуку, вставки та видалення.
- Використовується в багатьох бібліотеках, наприклад, в `std::map` і `std::set` в C++.

### b) **AVL-дерево (AVL Tree)**

**Основні характеристики:**
- Кожен вузол зберігає додаткову інформацію — фактор балансу (різниця висот лівого і правого піддерева).
- Висота дерева зберігається в межах O(log n).

**Переваги:**
- Підтримує баланс на кожному кроці, що забезпечує швидкий доступ до елементів.
- Операції пошуку, вставки і видалення виконуються за O(log n).

---

## 4. **Купа (Heap)**

**Основні характеристики:**
- Використовується для побудови пріоритетних черг.
- **Максимальна купа**: кожен батьківський елемент більше або рівне своїм нащадкам.
- **Мінімальна купа**: кожен батьківський елемент менше або рівне своїм нащадкам.

**Переваги:**
- Операція вставки та видалення елемента виконується за O(log n).
- Забезпечує швидкий доступ до максимального (або мінімального) елемента, що зручно для реалізації пріоритетних черг та алгоритмів сортування (Heap Sort).

---

## 5. **Трійкове дерево (Ternary Tree)**

**Основні характеристики:**
- Кожен вузол має до трьох нащадків.
  
**Переваги:**
- Може бути корисним у деяких спеціалізованих алгоритмах, наприклад, для певних видів пошукових задач.

---

## 6. **Б-бінарне дерево (B-Tree)**

**Основні характеристики:**
- Б-бінарне дерево — це багатовузлове дерево, де кожен вузол може мати більше ніж два нащадки.
- Кожен вузол має декілька ключів і дітей, що дозволяє ефективно працювати з великими наборами даних.

**Переваги:**
- Широко використовується для зберігання даних на диску, оскільки мінімізує кількість доступів до пам'яті.
- Підтримує швидкий пошук, вставку та видалення з часом O(log n).

---

## 7. **Тернарне дерево (Trie)**

**Основні характеристики:**
- Використовується для зберігання рядків.
- Кожен вузол може мати до 26 (або більше) дітей, які відповідають символам алфавіту.
  
**Переваги:**
- Операції пошуку, вставки та видалення можуть бути виконані за час, пропорційний довжині рядка.
- Дуже ефективно для задач, що потребують швидкого пошуку префіксів (наприклад, автозаповнення).

---

## 8. **Дерево К** (k-Tree)

**Основні характеристики:**
- Дерево, в якому кожен вузол може мати до **k** нащадків.

**Переваги:**
- Підходить для організації даних, коли кількість нащадків на рівні вузла більша за два (наприклад, дерева для зберігання ієрархій).

---

## Висновки:
- **Загальні дерева** (наприклад, бінарне дерево) добре підходять для загальних задач, де не вимагається строгий порядок елементів.
- **Збалансовані дерева** (AVL, червонопільні) забезпечують високу ефективність пошуку, вставки та видалення завдяки балансуванню.
- **Купи** є корисними для побудови пріоритетних черг.
- **Б-бінарні дерева** використовуються для великих структур даних, таких як бази даних, через їх здатність ефективно працювати з дисковими сховищами.



# № 4  Основні задачі, що вирішуються за допомогою дерев

## 1. **Пошук, вставка та видалення елементів** (Бінарне дерево пошуку)
- **Задача:** Потрібно ефективно зберігати і знаходити елементи.
- **Як вирішується:** Використовується **бінарне дерево пошуку (BST)** для виконання операцій пошуку, вставки та видалення за **O(log n)**.

---

## 2. **Сортування елементів** (Heap Sort)
- **Задача:** Потрібно відсортувати набір елементів.
- **Як вирішується:** Використовується **бінарна купа (Heap)** для сортування за **O(n log n)**.

---

## 3. **Пошук префіксів та автозаповнення** (Trie)
- **Задача:** Потрібно знайти або запропонувати можливі варіанти для введеного префікса (наприклад, автозаповнення в пошукових системах).
- **Як вирішується:** Використовується **дерево префіксів (Trie)** для ефективного пошуку за префіксом.

---

## 4. **Обчислення арифметичних виразів** (Бінарні дерева)
- **Задача:** Потрібно обчислити значення арифметичних виразів.
- **Як вирішується:** Використовується **бінарне дерево**, де кожен вузол — це операція, а листя — операнди.

---

## 5. **Генерація та перевірка ієрархій** (Організаційні структури, каталоги)
- **Задача:** Потрібно зберігати і обробляти багаторівневі ієрархії (наприклад, структури каталогів або організаційні діаграми).
- **Як вирішується:** Використовуються **дерева** для представлення таких ієрархій.

---

## Висновок:
Дерева є ефективними для вирішення задач, пов'язаних з пошуком, сортуванням, зберіганням і обробкою ієрархічних даних. Вони забезпечують швидке виконання операцій та є основою багатьох алгоритмів і структур даних.


# № 5 Організація купи (Heap)

Купа (Heap) — це спеціальна деревоподібна структура даних, яка задовольняє властивість купи:

- У **максимальній купі** кожен батьківський елемент не менший за своїх нащадків.
- У **мінімальній купі** кожен батьківський елемент не більший за своїх нащадків.

## Алгоритм додавання елемента в купу (Insert)

1. **Крок 1:** Додаємо новий елемент в останній доступний лист (правий нижній вузол дерева).
2. **Крок 2:** Порівнюємо доданий елемент з його батьком. Якщо елемент більший (для максимальної купи) — міняємо їх місцями.
3. **Крок 3:** Продовжуємо переміщення елемента до батька, поки не буде досягнута коренева вершина або поки елемент не стане на правильну позицію.

### Псевдокод для додавання елемента:

```python
def insert(heap, element):
    heap.append(element)  # Додаємо елемент в кінець масиву
    i = len(heap) - 1  # Індефікуємо індекс новододаного елемента
    while i > 0 and heap[i] > heap[(i - 1) // 2]:  # Поки елемент більше за батька
        heap[i], heap[(i - 1) // 2] = heap[(i - 1) // 2], heap[i]  # Міняємо місцями з батьком
        i = (i - 1) // 2  # Переходимо до батька


## Алгоритм вилучення елемента з купи (Delete)

1. **Крок 1:** Видаляємо корінь дерева (найбільший елемент у максимальній купі або найменший елемент у мінімальній купі).
2. **Крок 2:** Поміщаємо останній елемент з дерева на місце кореня.
3. **Крок 3:** Порівнюємо новий корінь з його дітьми.
4. **Крок 4:** Якщо корінь не задовольняє властивість купи, міняємо його місцями з найбільшим або найменшим нащадком.
5. **Крок 5:** Продовжуємо порівнювати нове місце елемента з його дітьми, поки не буде досягнута властивість купи або не буде досягнуто листа.

### Псевдокод для вилучення елемента:

```python
def delete(heap):
    if len(heap) == 0:
        return None  # Якщо купа порожня
    root = heap[0]  # Зберігаємо корінь
    heap[0] = heap[-1]  # Переміщаємо останній елемент в корінь
    heap.pop()  # Видаляємо останній елемент
    i = 0
    while 2 * i + 1 < len(heap):  # Поки є хоча б один нащадок
        left = 2 * i + 1
        right = 2 * i + 2
        larger = left  # Припускаємо, що лівий нащадок більший
        if right < len(heap) and heap[right] > heap[left]:  # Якщо правий нащадок більший
            larger = right
        if heap[i] >= heap[larger]:  # Якщо батько більший або рівний нащадку
            break
        heap[i], heap[larger] = heap[larger], heap[i]  # Міняємо місцями
        i = larger  # Переходимо до наступного рівня
    return root


## Порівняння часу виконання

- **Часова складність для додавання (Insert):** O(log n), де n — кількість елементів у купі.
- **Часова складність для вилучення (Delete):** O(log n), оскільки елемент може опуститися до самого низу дерева, і кожен рівень дерева містить не більше ніж O(log n) елементів.

## Висновок

Купа є ефективною структурою даних для реалізації пріоритетних черг, де операції вставки та вилучення виконуються за **O(log n)**. Вона також використовується в алгоритмах сортування, таких як **Heap Sort**, що забезпечує **O(n log n)** складність для сортування елементів.


# № 6 Задачі, які можна ефективно вирішити за допомогою купи

Купи є надзвичайно корисними для вирішення задач, де потрібно ефективно обробляти елементи за їх пріоритетом. Ось кілька прикладів:

## 1. Реалізація пріоритетної черги
- **Задача:** Операції вставки та вилучення елементів з черги з урахуванням пріоритету.
- **Приклад:** Система керування задачами, де задачі з вищим пріоритетом повинні виконуватися першими.
- **Часова складність:** Вставка та видалення — $ O(log n) $ .

## 2. Алгоритм сортування за допомогою купи (Heap Sort)
- **Задача:** Сортування елементів без використання додаткової пам'яті (на відміну від алгоритмів, що використовують додаткові структури даних).
- **Приклад:** Сортування масивів чи списків.
- **Часова складність:** $ O(n log n) $ .

## 3. Найменші/найбільші елементи в потоці даних
- **Задача:** Пошук k найбільших або найменших елементів у потоці даних.
- **Приклад:** Утримання поточної топ-10 новин або найбільших прибутків за певний період.
- **Часова складність:** Вставка нових елементів в купу — $ O(log k) $ , де k — кількість елементів у купі.

## 4. Алгоритм Дейкстри для пошуку найкоротших шляхів
- **Задача:** Пошук найкоротшого шляху від однієї вершини до всіх інших у графі з невід'ємними вагами ребер.
- **Приклад:** Використання в системах GPS для пошуку найкоротшого шляху.
- **Часова складність:** O((V + E) log V), де V — кількість вершин, E — кількість ребер.

## 5. Алгоритм Кушала для обробки потоків даних
- **Задача:** Постійний моніторинг та виведення k найбільших елементів у великому потоці.
- **Приклад:** Використовується в соціальних мережах для визначення найпопулярніших постів або твітів.
- **Часова складність:** O(log k) для кожної вставки.

## 6. Керування пам'яттю
- **Задача:** Використання купи для ефективного управління пам'яттю, де найбільші блоки пам'яті вилучаються першими.
- **Приклад:** Система управління пам'яттю в операційних системах.
- **Часова складність:** O(log n) для видалення найбільш вільного блоку.

## 7. Задача з K найбільшими елементами (Top K elements)
- **Задача:** Потрібно знайти K найбільших елементів серед величезної кількості чисел.
- **Приклад:** Пошук топ-5 найбільших покупок за певний період.
- **Часова складність:** O(n log k), де n — кількість елементів, k — кількість найбільших елементів.

---

## Висновок
Купи є надзвичайно потужною структурою даних, коли необхідно швидко знаходити елементи за пріоритетом. Вони застосовуються в таких алгоритмах, як Heap Sort, алгоритм Дейкстри, а також для вирішення задач з потоками даних або пріоритетними чергами. Часова складність основних операцій (вставка, вилучення) в купі — O(log n), що робить їх ефективними для багатьох задач.


# № 7 Як геш-функція використовується для зберігання та пошуку даних в хеш-таблиці?

Геш-функція є важливою частиною механізму роботи хеш-таблиці, яка дозволяє ефективно зберігати та знаходити дані. Ось основні етапи використання геш-функції:

## 1. Зберігання даних в хеш-таблиці

- **Крок 1:** Геш-функція приймає ключ (наприклад, рядок, ціле число або інший тип даних) і генерує індекс або хеш-значення. Це значення визначає місце, де дані будуть збережені в хеш-таблиці.
  
  - Наприклад, для ключа "apple" геш-функція може обчислити хеш-значення 5, що означатиме, що елемент з таким ключем буде збережений на індекс 5 хеш-таблиці.
  
- **Крок 2:** Дані (значення) зберігаються в цьому місці таблиці. Якщо кілька ключів мають однакове хеш-значення (колізія), використовуються методи вирішення колізій (наприклад, ланцюгове хешування або відкриті адресації).

## 2. Пошук даних у хеш-таблиці

- **Крок 1:** Для пошуку елемента за певним ключем, спочатку обчислюється хеш-значення цього ключа за допомогою геш-функції.
  
  - Наприклад, якщо потрібно знайти елемент з ключем "apple", геш-функція обчислює хеш-значення для цього ключа.
  
- **Крок 2:** Використовуючи отримане хеш-значення, шукається потрібний елемент у хеш-таблиці. Якщо на цьому індексі є елемент з потрібним ключем, то повертається його значення.
  
  - Якщо кілька елементів мають однаковий хеш (колізія), застосовуються методи вирішення колізій (наприклад, перевірка всіх елементів у зв'язку або наступних індексах в разі відкритої адресації).

## Важливість геш-функції

- **Ефективність пошуку:** Геш-функція дозволяє значно скоротити час пошуку елемента. Без геш-функції пошук елемента вимагав би перебору всіх елементів у таблиці (O(n)), в той час як з геш-функцією пошук зазвичай відбувається за O(1) в середньому випадку.
  
- **Розподіл елементів:** Хороша геш-функція повинна рівномірно розподіляти ключі по хеш-таблиці, щоб знизити кількість колізій і забезпечити високу ефективність.

## Проблеми та вирішення

- **Колізії:** Коли два або більше ключів дають однаковий хеш-індекс, виникає колізія. Для вирішення колізій застосовуються методи:
  - **Ланцюгове хешування:** Кожен індекс таблиці містить список (ланцюг) елементів, які мають однаковий хеш.
  - **Відкрита адресація:** У разі колізії шукається наступне вільне місце в таблиці для зберігання елемента.

## Висновок

Геш-функція є основою ефективного зберігання та пошуку даних в хеш-таблиці. Вона дозволяє швидко знаходити елементи за ключем, оскільки переведення ключа у хеш-значення забезпечує доступ до конкретного індексу в таблиці. Водночас важливо мати хорошу геш-функцію для уникнення колізій та забезпечення ефективної роботи хеш-таблиці.


# № 8 Методи вирішення колізій в хеш-таблицях

Колізії в хеш-таблиці виникають, коли два або більше ключі мають однакове хеш-значення. Для того, щоб зберігати всі елементи, необхідно застосовувати методи вирішення колізій. Існують два основні підходи:

## 1. Ланцюгове хешування (Chaining)

### Опис:
- Кожен індекс хеш-таблиці містить зв'язаний список (або іншу структуру даних), в якому зберігаються всі елементи з однаковим хеш-значенням.
- При колізії новий елемент додається до списку на відповідному індексі.

### Переваги:
- **Простота реалізації:** Ланцюгове хешування легко реалізується та дозволяє зберігати кілька елементів на одному індексі без необхідності переміщення інших елементів.
- **Гнучкість:** Можна зберігати будь-яку кількість елементів на кожному індексі, оскільки не обмежує розмір таблиці.
- **Зменшення кількості колізій:** Навіть при великій кількості колізій швидкість доступу до елементів залишається високою, оскільки в найгіршому випадку шукається лише кілька елементів в списку.

### Недоліки:
- **Збільшення використання пам'яті:** Для кожного елемента необхідно додавати додаткову пам'ять для збереження зв'язків між елементами.
- **Продуктивність при великій кількості колізій:** Якщо розмір таблиці обмежений і дуже багато елементів потрапляють в одну корзину, ефективність пошуку може знизитися до O(n) для кожного пошуку, де n — кількість елементів у ланцюзі.

---

## 2. Відкрита адресація (Open Addressing)

### Опис:
- У випадку колізії шукається інший вільний індекс у хеш-таблиці для зберігання елемента.
- Існує кілька варіантів відкритої адресації:
  - **Лінійне пробування:** Перевірка наступних індексів $ (i+1, i+2, ...) $ до знаходження вільного.
  - **Квадратичне пробування:** Перевірка індексів $ (i+1^2, i+2^2, i+3^2, ...) $ .
  - **Хешування подвійними хеш-функціями:** Використання іншої хеш-функції для пошуку наступного вільного індексу.

### Переваги:
- **Менше використання пам'яті:** Відкрита адресація не потребує додаткової пам'яті для зберігання зв'язків, оскільки всі елементи зберігаються безпосередньо в масиві.
- **Простота реалізації:** Відкрите адресування зазвичай простіше реалізувати, ніж ланцюгове хешування, оскільки не потрібно управляти вказівниками або списками.

### Недоліки:
- **Проблеми з ефективністю при заповненості:** Коли таблиця заповнена більше ніж на 70%, ефективність операцій знижується, оскільки потрібно більше часу для пошуку вільних місць.
- **Повільний пошук:** Якщо багато елементів мають однаковий хеш, пошук вільного місця може займати багато часу (до $ O(n)) $ .
- **Ускладнене видалення:** При видаленні елементів може бути складно коректно відновити таблицю, оскільки пробування наступних індексів може порушити структуру.

---

## 3. Динамічне хешування

### Опис:
- У цей метод включаються додаткові стратегії, такі як зміна розміру хеш-таблиці при досягненні певного рівня заповнення, а також використання багаторівневих таблиць.
- Цей метод дозволяє уникати проблем з перевантаженням таблиці, збільшуючи її розмір в разі необхідності.

### Переваги:
- **Автоматичне масштабування:** Виведення таблиці за межі її початкового розміру дозволяє ефективно працювати з великими наборами даних.
- **Добре вирішує проблему колізій при великій кількості даних.**

### Недоліки:
- **Вартість перерахунку:** Розширення таблиці вимагає перерахунку всіх елементів, що може бути витратним по часу, особливо при великих таблицях.
- **Складність реалізації:** Реалізація динамічного хешування є більш складною, ніж інші методи.

---

## Висновок

Існують різні методи вирішення колізій у хеш-таблицях, і кожен з них має свої переваги та недоліки:
- **Ланцюгове хешування** є гнучким та простим у використанні, але може вимагати додаткової пам'яті.
- **Відкрита адресація** дозволяє економити пам'ять, але може зіткнутися з проблемами ефективності при заповненості таблиці.
- **Динамічне хешування** дозволяє працювати з великими об'ємами даних, але є складним у реалізації та потребує додаткових витрат на перерахунок таблиці.
